controllers/pvc.go

```diff
@@ -1,9 +1,155 @@
 package controllers
 
 import (
+	"errors"
+	"fmt"
+
 	"github.com/go-logr/logr"
+	pvcv1alpha1 "github.com/konveyor/volume-snapshot-mover/api/v1alpha1"
+	snapv1 "github.com/kubernetes-csi/external-snapshotter/client/v4/apis/volumesnapshot/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
 func (r *DataMoverBackupReconciler) BindPVC(log logr.Logger) (bool, error) {
+	// Get datamoverbackup from cluster
+	dmb := pvcv1alpha1.DataMoverBackup{}
+	if err := r.Get(r.Context, r.NamespacedName, &dmb); err != nil {
+		return false, err
+	}
+	// Check if Volumesnapshot is present in the protected namespace
+	vs := snapv1.VolumeSnapshot{}
+	if err := r.Get(r.Context,
+		types.NamespacedName{Name: fmt.Sprintf("%s-volumesnapshot", dmb.Spec.VolumeSnapshotContent.Name), Namespace: r.NamespacedName.Namespace}, &vs); err != nil {
+		return false, errors.New("cloned volumesnapshot not available in the protected namespace")
+	}
+
+	// Create a PVC with the above volumesnapshot as the source
+	pvc := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-pvc", dmb.Spec.VolumeSnapshotContent.Name),
+			Namespace: r.NamespacedName.Namespace,
+		},
+	}
+
+	// check if pvc exists
+	// create pvc only if it doesnt exists
+	_ = r.Get(r.Context, r.NamespacedName, pvc)
+
+	if pvc.Status.Phase == "" {
```

Comment:
@dymurray: I don't think using `CreateOrUpdate` on PVC is the right call since the spec field is immutable. Should I switch to a different option?

Reply:
@savitharaghunathan In the `buildPVC` function if you try to update each PVC spec field individually I believe you should be able to mitigate the immutable field issue. For example:
```
pvc.Spec.DataSource = &corev1.TypedLocalObjectReference{
			Name:     vs.Name,
			Kind:     vs.Kind,
			APIGroup: &vs.APIVersion,
		}

pvc.Spec.AccessModes = []corev1.PersistentVolumeAccessMode{
			"ReadWriteOnce",
		},
...
```
Reference OADP registry service: https://github.com/openshift/oadp-operator/blob/a4ede2a4d892d3d45cc87ca7ed8ec0f05a4edb11/controllers/registry.go#L848

Reply:
Thanks @shubham-pampattiwar :) I saw this error with respect to `StorageClassName` field. I am not setting it explicitly in the spec, so when I ran the project again that where it complained. 
`error: PersistentVolumeClaim "snapcontent-365946e1-66b6-4e5b-8e3f-b92e6b72f441-pvc" is invalid: spec: Forbidden: spec is immutable after creation except resources.requests for bound claims
  core.PersistentVolumeClaimSpec{
  	... // 2 identical fields
  	Resources:        {Requests: {s"storage": {i: {...}, s: "10Gi", Format: "BinarySI"}}},
  	VolumeName:       "",
- 	StorageClassName: nil,
+ 	StorageClassName: &"gp2-csi",
  	VolumeMode:       &"Filesystem",
  	DataSource:       nil,
  }`
  
  Unless I specify the `StorageClassName` in the spec, I am not sure if the issue will be resolved. On that note, I have no idea where to get the `StorageClassName` value from. I left it empty so that the PVC could pickup the default one. Should I make it as a field in DMB CR?

Reply:
Yeah you could make that as a field in DMB CR or if you just want to use the default storage class, you can try to list the available storage classes with criteria that the storage class has the annotation `storageclass.kubernetes.io/is-default-class: "true"` implying that its the default one and pass that as a function parameter to the `buildPVC` function and assign to the PVC spec. WDYT ?

Reply:
@shubham-pampattiwar just keep in mind that there may not be a default storage class. Also, it's possible for more than one storage classes to be labeled default, but if that happens, this should be considered equivalent to having no default storage class.